### DISCUSSION TITLE
Add support for deferred column loading

### DISCUSSION THREAD

**Discussion by hitman-gdg** [fastapi/sqlmodel#1931] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from sqlmodel import deferred

class User(sqlmodel, table=True):
    id: Optional[int] = Field(primary_key=True, default=None)
    name: str
    data: Optional[bytes] = Field(deferred=True)
```

### Description

The ability to be able to included deferred sql columns

### Wanted Solution

I would like to be able to work with sqlalchemy deferred columns, so that certain columns are not loaded by default as part of a model

### Wanted Code

```python
from sqlmodel import deferred
```

### Operating System

macOS

### SQLModel Version

0.0.4

### Python Version

3.9.6

**Comment by eddcorts** [comment]
Having same issue, did you find any alternative with base SQLAlchemy or something?

**Comment by FilipeMarch** [comment]
Did you find a solution?

**Comment by lucas-labs** [comment]
I think that we can use the [new `sa_type`](https://github.com/tiangolo/sqlmodel/pull/505) now, in order to achieve this behaviour.

**Comment by ngi** [comment]
any news how to use sa_type for deferred loading?

**Comment by MarkusDressel** [comment]
@tiangolo can you provide some information if / how it possible to leverage SQLalchemy's deferred loading mechanism in SQLModel ?

I also could not find a way to make deferred loading on column level (not on query level) work. sa_column setting is not recognized by SQLModel and still loads the deferred column
```python
blob: bytes | None = Field(default=None, sa_column=deferred(Column(LargeBinary)))
```

**Comment by Fatem-Mq** [comment]
Any updates? =(

**Comment by shawnmartin-py** [comment]
Why anyone would even consider sqlmodel these days is beyond me.
Been fine managing Pydantic and SQLA models independently for the past 3 years.

**Reply by riotsnotdiets** [reply]
I am starting to think the same :(

**Comment by shawnmartin-py** [comment]
It's a lib that get's hardly any updates, and only 1 person is allowed to merge PR's. 🤦‍♂️

**Comment by alessiamarcolini** [comment]
Hi! Is there an update on how to make this work? :) Thanks.

**Comment by YuriiMotov** [comment]
It seems that for now there is no way to declare model with the deferred field, and we can only defer column loading by specifying the option for query:

```py
session.get(User, 1, options=[defer(User.data)])
```

or

```py
u = session.exec(
    select(User).where(col(User.id) == 1).options(defer(User.data))
).one()
```

**Comment by justanotherbody** [comment]
Came looking for support, found this issue, determined a workaround...


```python
from typing import List

import sqlalchemy
import sqlalchemy.orm

import sqlmodel


def defer_columns(*columns: List[str]):
    def defer_columns_for_sqlmodel_class(cls):
        mapper = sqlalchemy.inspect(cls)
        for column in columns:
            prop = mapper.get_property(column)
            prop.deferred = True
            new_strategy = []
            # we expect strategy_key to be a tuple of (name, value) pairs but we're
            # going to be extra careful here...
            for strategy in prop.strategy_key:
                if (
                    isinstance(strategy, tuple)
                    and len(strategy) == 2
                    and isinstance(strategy[0], str)
                    and strategy[0] == "deferred"
                ):
                    new_strategy.append(("deferred", True))
                else:
                    new_strategy.append(strategy)
            prop.strategy_key = tuple(new_strategy)

        # normally invoked automatically, but you might need to do this explicitly if
        # you're up to really unusual things...
        # sqlalchemy.orm.configure_mappers()

        return cls

    return defer_columns_for_sqlmodel_class


@defer_columns("deferred_by_default")
class Foo(sqlmodel.SQLModel, table=True):

    id: int = sqlmodel.Field(primary_key=True)
    always_loaded: str
    deferred_by_default: str


def get_attrs(model):
    "Return underling attributes."
    return model._sa_instance_state.dict


# test_engine is a pytest fixture with a working test DB connection. YMMV
def test_deferred(test_engine):
    session = sqlmodel.Session(test_engine)

    session.add(Foo(id=1, always_loaded="Happy", deferred_by_default="Sad"))
    session.commit()
    session.expire_all()

    select = sqlmodel.select(Foo).where(Foo.id == 1)

    explicitly_deferred = session.exec(
        select.options(sqlalchemy.orm.defer(Foo.deferred_by_default))
    ).one()

    assert "always_loaded" in get_attrs(explicitly_deferred)
    assert "deferred_by_default" not in get_attrs(explicitly_deferred)
    session.expire_all()

    explicitly_loaded = session.exec(
        select.options(sqlalchemy.orm.undefer(Foo.deferred_by_default))
    ).one()
    assert "deferred_by_default" in get_attrs(explicitly_loaded)
    session.expire_all()

    foo = session.exec(select).one()
    assert "deferred_by_default" not in get_attrs(foo)
```

**Comment by natikos** [comment]
any updates on this issue?

**Comment by riotsnotdiets** [comment]
Any updates on the issue?
